gadget/ondisk.go

```diff
@@ -130,21 +126,31 @@ func fromSfdiskPartitionType(st string, sfdiskLabel string) (string, error) {
 	}
 }
 
-func blockDeviceSizeInSectors(devpath string) (quantity.Size, error) {
-	// the size is reported in 512-byte sectors
-	// XXX: consider using /sys/block/<dev>/size directly
-	out, err := exec.Command("blockdev", "--getsz", devpath).CombinedOutput()
+func blockdevCmd(cmd, devpath string) (quantity.Size, error) {
+	out, err := exec.Command("blockdev", cmd, devpath).CombinedOutput()
 	if err != nil {
 		return 0, osutil.OutputErr(out, err)
 	}
 	nospace := strings.TrimSpace(string(out))
 	sz, err := strconv.Atoi(nospace)
 	if err != nil {
-		return 0, fmt.Errorf("cannot parse device size %q: %v", nospace, err)
+		return 0, fmt.Errorf("cannot parse blockdev size %q: %v", nospace, err)
 	}
 	return quantity.Size(sz), nil
 }
 
+func blockDeviceSizeInSectors(devpath string) (quantity.Size, error) {
+	// the size is always reported in 512-byte sectors, even if the device does
+	// not have a physical sector size of 512
+	// XXX: consider using /sys/block/<dev>/size directly
+	return blockdevCmd("--getsz", devpath)
+}
+
+func blockDeviceSectorSize(devpath string) (quantity.Size, error) {
+	// the size is reported in raw bytes
+	return blockdevCmd("--getbsz", devpath)
```

Comment:
this should actually be

```suggestion
	return blockdevCmd("--getpbsz", devpath)
```

which returns the real logical sector size (which the blockdev man page confusing calls "physical"), see on a physical Intel NUC with UC20:

```
ubuntu@dawson-005-201805-26255:~$ sudo blockdev --getpbsz /dev/sda
512
ubuntu@dawson-005-201805-26255:~$ sudo blockdev --getbsz /dev/sda
4096
ubuntu@dawson-005-201805-26255:~$ sudo sfdisk --json /dev/sda
{
   "partitiontable": {
      "label":"gpt",
      "id":"C5C04F0C-1853-469E-8B6F-922B68203368",
      "device":"/dev/sda",
      "unit":"sectors",
      "firstlba":34,
      "lastlba":468862094,
      "partitions": [
         {"node":"/dev/sda1", "start":2048, "size":2048, "type":"21686148-6449-6E6F-744E-656564454649", "uuid":"71D1744C-1EBC-4CFA-A427-BC4B4F7AA6F6", "name":"BIOS Boot"},
         {"node":"/dev/sda2", "start":4096, "size":2457600, "type":"C12A7328-F81F-11D2-BA4B-00A0C93EC93B", "uuid":"DC9074F5-43F5-4D5F-A8D0-F82EE2C16C7E", "name":"ubuntu-seed"},
         {"node":"/dev/sda3", "start":2461696, "size":1536000, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"6FD0E472-6970-7147-8A79-72704CA1DDEE", "name":"ubuntu-boot"},
         {"node":"/dev/sda4", "start":3997696, "size":32768, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"2F59710C-9111-CA4C-BD3C-323911391C79", "name":"ubuntu-save"},
         {"node":"/dev/sda5", "start":4030464, "size":464831631, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"D0A84F6F-32CF-F540-9EF9-9ACFF27F0EBE", "name":"ubuntu-data"}
      ]
   }
}
```

and on a VM:

```
root@ubuntu:/# blockdev --getpbsz /dev/vda
512
root@ubuntu:/# blockdev --getbsz /dev/vda
4096
root@ubuntu:/# sfdisk --json /dev/vda
{
   "partitiontable": {
      "label":"gpt",
      "id":"4646D35B-B736-4140-B42B-E9DB85D71C0E",
      "device":"/dev/vda",
      "unit":"sectors",
      "firstlba":34,
      "lastlba":6291422,
      "partitions": [
         {"node":"/dev/vda1", "start":2048, "size":2048, "type":"21686148-6449-6E6F-744E-656564454649", "uuid":"0A156DB8-7E87-4CBE-8C8E-7075452D0D23", "name":"BIOS Boot"},
         {"node":"/dev/vda2", "start":4096, "size":2457600, "type":"C12A7328-F81F-11D2-BA4B-00A0C93EC93B", "uuid":"FA96BADF-99CF-44A3-A925-2AA916C29DD8", "name":"ubuntu-seed"}
      ]
   }
}
```

and on my 4K disk in the dragonboard:

```
test-ubuntu@ubuntu:~$ sudo blockdev --getbsz /dev/sda
4096
test-ubuntu@ubuntu:~$ sudo blockdev --getpbsz /dev/sda
4096
test-ubuntu@ubuntu:~$ sudo sfdisk --dump --json /dev/sda
{
   "partitiontable": {
      "label":"gpt",
      "id":"A671BE0F-C1E7-EEB5-1E50-BB71D4D075BB",
      "device":"/dev/sda",
      "unit":"sectors",
      "firstlba":6,
      "lastlba":6335454,
      "table-length":"32",
      "partitions": [
         {"node":"/dev/sda1", "start":256, "size":2560, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"07EC888A-9D08-6A47-B90D-04FF3CC926B2", "name":"ubuntu-save"},
         {"node":"/dev/sda2", "start":2816, "size":2560, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"90BC94E9-6DFF-FA43-B4B3-3BCCBCCB7508", "name":"ubuntu-boot"},
         {"node":"/dev/sda3", "start":5376, "size":128000, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"D2C6D097-928F-D7BD-05F0-3BE7488C8157", "name":"ubuntu-seed"},
         {"node":"/dev/sda4", "start":133376, "size":6202079, "type":"0FC63DAF-8483-4772-8E79-3D69D8477DE4", "uuid":"E39E7A6A-6516-5346-A277-13A2B5FECEFA", "name":"ubuntu-data"}
      ]
   }
}

```